File: 66.js
/**
 * @param {number[]} digits
 * @return {number[]}
 */
var plusOne = function (digits) {
    let j = digits.length - 1;
    while (digits[j] === 9 && j >= 0) {
        j--;
    }
    if (j >= 0) {
        digits[j] += 1;
    }
    else {
        digits[0] = 1;
        digits.push(0);
        j = 0;
    }

    j++;
    while (j < digits.length) {
        digits[j] = 0;
        j++;
    }

    return digits;
};

console.log(plusOne([1, 2, 3]))
console.log(plusOne([9,9,9]))
console.log(plusOne([1, 9, 9]))

File: cpp/66.cpp
#include <iostream>
using namespace std;

class Solution1 {
public:
    vector<int> plusOne(vector<int>& digits) {
        int i = digits.size() - 1;
        while(i >= 0 && digits[i] == 9) {
            digits[i] = 0;
            i--;
        }
        if(i >= 0) {
            digits[i]++;
        } 
        else {
            digits.insert(digits.begin(), 1);
        }
        return digits;
    }
};

class Solution {
public:
    vector<int> plusOne(vector<int>& digits) {
        int i = digits.size() - 1;
        vector<int> res;
        while(i >= 0 && digits[i] == 9) {
            digits[i] = 0;
            i--;
        }
        if(i >= 0) {
            digits[i]++;
        }
        else {
            res.push_back(1);
        }
        for(int digit : digits) 
            res.push_back(digit);
        return res;
    }
};


int main() {

    Solution solution;

    cout << solution.() << endl;

    return 0;
}